### PR TITLE
Add pull request review guidelines

### PR DESCRIPTION
**Pull request by clayjohn**
This supercedes: https://github.com/godotengine/godot-docs/pull/5191

I opened a new PR to make it a bit easier to track the conversation as the conversion in #5191 is getting a little out of date. 

In particular, this version expands more on our greater emphasis on style review and on the production team that we formed this year. 

CC @YuriSizov @akien-mga @mhilbrunner

### CODE REVIEW

File: community/contributing/pr_review_guidelines.rst
Comment (mhilbrunner) [inline community/contributing/pr_review_guidelines.rst:15]: "..(optimally closing Github issue or implementing a proposal)", could also link to more info about the proposal process?

File: community/contributing/pr_review_guidelines.rst
Comment (mhilbrunner) [inline community/contributing/pr_review_guidelines.rst:26]: This can link to the teams page.

File: community/contributing/pr_review_guidelines.rst
Comment (mhilbrunner) [inline community/contributing/pr_review_guidelines.rst:46]: Nit: Has 'typically' twice in 2 sentences. Maybe replace one with 'often' or 'most of the time' or similar.

File: community/contributing/pr_review_guidelines.rst
Comment (mhilbrunner) [inline community/contributing/pr_review_guidelines.rst:87]:
```suggestion
When reviewing a pull request, ensure that nothing outside of the PR target area
```
I think?

**Review by mhilbrunner — COMMENTED** [review]
Great work! As discussed in the docs meeting, this needs Rémi's final approval.

Maybe we could have a section dedicated to what _not_ to do when reviewing PRs?
Some thoughts:
- Try to avoid needless double reviews, i.e. if you can, review the full PR, avoiding the thing where you note the first issue, comment that, the contributor fixes and commits and you spot another issue elsewhere, ad nauseam. Of course, can't always be avoided, but we should try not to.
- Try to avoid being overly nitpicky. This is a delicate balance, but I've seen PRs that are clear improvements get stale and abandoned because it could still be even better, although the existing code is way worse. The perfect is the enemy of the good, and all that.
- In the same vein, try to avoid throwing more work at the contributor. Providing context or related/similar issues or proposals that may be fixed similarly is good, but adding a "may as well fix that thing over there as well while at it" isn't fair to the contributor. Reviews should stay in scope of the PR.
- If something is a no go/has clear blockers/will likely not get merged/its doubtful and needs more consensus/whatever should be communicated as early and clearly as possible. Way better than stringing people along because it feels bad to say "sorry, no".
- Try to avoid asking for needless rebases. PRs can be rebased once its clear they are approved/mergeable, or if they really need to be tested with a more recent change.
- We should link to the code of conduct for review, i.e., assume the best, try to be polite.

**Comment by clayjohn** [comment]
> 

Thanks for the feedback! I have incorporated all your suggestions into an updated draft.

**Comment by skyace65** [comment]
Should there be anything on the page for reviewing documentation? Or would that be a big enough section that it should be made in a later PR?

**Comment by YuriSizov** [comment]
@skyace65 I think a lot of it already applies to the documentation as well. The correctness of the information should be validated by the corresponding maintainer team/developer, and the language and formatting should be validated by the documentation maintainers (akin to how we check for style in the code PRs).

We can link to the part of the online documentation about contributing docs, specifically for writing style and whatnot.

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:82]:
```suggestion
2. Test the PR and look for regressions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Formatting issue.

Also, in your rework the part about testing the PR for its proposed functionality is kind of amiss. It was point 3 in my PR, but here point 3 is different, and I can't find any similar description.

I think we should still incorporate the message from my PR about testing the improved functionality here, in some form. Either as a part of this point 2, or in place of your point 3 (which I've already proposed to move elsewhere, so it all fits together nicely).

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:149]: Should link directly to the section, I think. You'd need to add `.. _doc_something:` before the section title to be able to link to it.

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:197]:
```suggestion
whether you reviewed for style as well as correctness, etc.). Even if you are not
```

Missing comma, I think?

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:210]: "code style of code" probably needs one less "code", the first one I think.

```suggestion
other words, maintainers shouldn't ask contributors to change the style of
code that may need to be rewritten in subsequent reviews. Similarly, maintainers
```

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:225]: I think we can merge these two paragraph. This section looks a bit too broken up into way too small paragraphs, and here's a good opportunity to unite some of these messages, as these lines relate.

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:227]:
```suggestion
For example, you should check that:
```

A bit more direction here helps, IMO.

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:251]: I noticed that sometimes "In general" has a comma, and sometimes it doesn't. Probably should be consistent throughout the doc?

I would also change this particular bit to "In practice" and I would break it into a separate paragraph so it's easier to notice. Otherwise it gets lost in the middle of a rather long passage.

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:296]: Formatting is broken here. I think you need to add a new line before the list. Otherwise it doesn't get parsed into one, and is just appended as plain text to the preceding paragraph.

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:317]: Worth adding that if a PR has been around for some time, it may need a rebase even if GitHub doesn't report any conflicts. This may be due to the diff tool being unstable in some cases when lines close to each other get modified, or because we've changed the CI to catch more issues.

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:322]: I think we capitalized it as "GitHub", by the way. It should be used like that consistently throughout this article (it's mostly "Github" now as far as I can see).

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:328]: And I think we capitalized "Git" as well.

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:337]: Apparently, we have a good section on this in the CONTRIBUTING.md document in the main repo, we can link to it, or at least still an example from it:

https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#format-your-commit-messages-with-readability-in-mind

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:364]: Should also mention that sometimes PRs will have a wildcard milestone assigned to them, indicating that they can be merged in any upcoming release matching that wildcard. When merging such PR the milestone needs to be updated to the actual currently-in-development version.

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:395]: Did you intend for this sentence to be incomplete? As in, should it be read as "Merge the pull request, if ...", together with the title?

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:396]: I know this was a big point of this document, but as a closing sentence it looks a bit weird. Maybe "... you can go ahead and merge ..." instead of "don't be afraid"?

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:247]: Can't help but notice that we don't have any mention for the fact that you shouldn't merge your own pull requests. I think it's important to spell out, even though we repeat a few times that merging is done by the production team or team leaders (a team leader shouldn't merge their own PRs either, normally).

**Review by YuriSizov — CHANGES_REQUESTED** [review]
I like it a lot! I think you took my original article and made it look much more professional.

Overall, I'd even be okay to merge it as is, but I have some suggestions and corrections. Namely, a couple of, IMO, important points are missing that could be added.

Of the things I didn't make a dedicated review comment about, I was tripped by the word "substantive" every time it was used. I understand it, but it might be a bit too dry and bureaucratic, perhaps? Like a jargon you'd find in an official document rather than an engine documentation. But up to you on this one.

**Comment by YuriSizov** [comment]
As an extra, it occurred to me today that we can also link to a filter like this in the "Merging" section:
https://github.com/godotengine/godot/pulls?q=is%3Apr+is%3Aopen+review%3Aapproved+sort%3Aupdated-desc

It can come with a tip that if you're looking for something potentially ready to be merged, check this list.

File: community/contributing/pr_review_guidelines.rst
Comment (clayjohn) [reply community/contributing/pr_review_guidelines.rst:99]: I don't know, the intention here was not to provide a step by step guide, but to provide a list of things that you should do in reviewing a PR. The idea is that the most important tips should have their own header/section so they are very visible. Moving this into section 5 will bury it

File: community/contributing/pr_review_guidelines.rst
Comment (clayjohn) [reply community/contributing/pr_review_guidelines.rst:82]: I removed the previous section 3 intentionally as I didn't feel it added much value on top of the other sections (same thing with section 0).

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [reply community/contributing/pr_review_guidelines.rst:99]: Perhaps, but the rest of this list reads quite naturally as a plan that you can follow. This point just looks out of place in the middle. Can also be made point 1, if you think it's important to keep it as a separate point. Can also be made a notice block in point 5 to still keep it visible. IMO it ties into the idea behind point 5 for that, but it's fine either way.

File: community/contributing/pr_review_guidelines.rst
Comment (clayjohn) [reply community/contributing/pr_review_guidelines.rst:247]: Its mentioned under ``Get feedback from the right people/teams``. I went with a pretty soft approach on this as right now many PRs are merged and reviewed by Akien alone. We can strengthen the language once we formalize our practice a bit more. As it is right now we would open ourselves to criticism if we use strong language

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [reply community/contributing/pr_review_guidelines.rst:82]: I get that testing the feature itself is kind of expected, but I think it would be good to cover even the obvious parts in this document. At least in passing, as we are making a checklist for maintainers to follow.

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [reply community/contributing/pr_review_guidelines.rst:247]: That section talks about not merging PRs based on your own sole review. I'm talking about merging a PR that you yourself have authored. It was point 0 in one of the sections of my original PR, if that helps.

Even Akien doesn't like to merge his own PRs, but there is often little help with some stuff that he changes. It would be nice to underline that self-merges are not expected nor desired. Even if to show intent.

File: community/contributing/pr_review_guidelines.rst
Comment (clayjohn) [reply community/contributing/pr_review_guidelines.rst:82]: Okay, added a note about ensuring the PR does what it does it does (fixes bug or implements feature)

File: community/contributing/pr_review_guidelines.rst
Comment (YuriSizov) [inline community/contributing/pr_review_guidelines.rst:19]:
```suggestion
  1. A contributor opens a PR that fixes a specific problem (optimally closing
    a GitHub `issue <https://github.com/godotengine/godot>`_ or implementing
    a `proposal <https://github.com/godotengine/godot-proposals>`_).
```

I think you've lost a couple of words here, "optimally closing a", during editing.

**Comment by mhilbrunner** [comment]
This needs a rebase after https://github.com/godotengine/godot-docs/pull/6516 - also whats the state on this? :)

**Comment by clayjohn** [comment]
Yuri and I have discussed a few changes I need to make, but ultimately we need @akien-mga to also review this before moving forward

**Comment by YuriSizov** [comment]
Rebased and resolved the remaining suggestions and corrections.

**Review by YuriSizov — APPROVED** [review]
Given that this is a work of 2 production team members (Clay and me), it was positively reviewed by Max, and Remi was already happy with the content before the last rework (asking only to improve the structure), I think this is good to merge now. We can always iterate on it later.

**Comment by YuriSizov** [comment]
Thanks, everyone!

**Comment by mhilbrunner** [comment]
Pushed to `4.0` and `stable`.

**Pull request merged**
